File: frontend/src/types/project.ts
// Project management module types for admin panel enhancement

export interface Task {
  id: string;
  title: string;
  description?: string;
  type: 'feature' | 'bug' | 'improvement' | 'maintenance' | 'research' | 'documentation';
  status: 'backlog' | 'todo' | 'in_progress' | 'review' | 'testing' | 'done' | 'cancelled';
  priority: 'low' | 'medium' | 'high' | 'critical';
  assigneeId?: string;
  assigneeName?: string;
  reporterId: string;
  reporterName: string;
  projectId?: string;
  milestoneId?: string;
  sprintId?: string;
  labels: string[];
  estimatedHours?: number;
  actualHours?: number;
  storyPoints?: number;
  dueDate?: Date;
  startDate?: Date;
  completedDate?: Date;
  createdAt: Date;
  updatedAt: Date;
  comments: TaskComment[];
  attachments: TaskAttachment[];
  dependencies: TaskDependency[];
  subtasks: Task[];
  parentTaskId?: string;
}

export interface TaskComment {
  id: string;
  taskId: string;
  authorId: string;
  authorName: string;
  content: string;
  createdAt: Date;
  updatedAt?: Date;
  isEdited: boolean;
}

export interface TaskAttachment {
  id: string;
  taskId: string;
  fileName: string;
  fileSize: number;
  fileType: string;
  url: string;
  uploadedBy: string;
  uploadedAt: Date;
}

export interface TaskDependency {
  id: string;
  taskId: string;
  dependsOnTaskId: string;
  type: 'blocks' | 'relates_to' | 'duplicates' | 'caused_by';
  createdAt: Date;
}

export interface Milestone {
  id: string;
  name: string;
  description?: string;
  projectId?: string;
  status: 'planning' | 'active' | 'completed' | 'cancelled' | 'overdue';
  startDate: Date;
  dueDate: Date;
  completedDate?: Date;
  progress: number; // 0-100
  totalTasks: number;
  completedTasks: number;
  totalStoryPoints?: number;
  completedStoryPoints?: number;
  budget?: number;
  spentBudget?: number;
  currency?: string;
  ownerId: string;
  ownerName: string;
  createdAt: Date;
  updatedAt: Date;
  tasks: Task[];
}

export interface Sprint {
  id: string;
  name: string;
  goal?: string;
  projectId?: string;
  status: 'planning' | 'active' | 'completed' | 'cancelled';
  startDate: Date;
  endDate: Date;
  capacity: number; // story points or hours
  commitment: number; // planned story points or hours
  completed: number; // actual completed story points or hours
  velocity: number; // average velocity from previous sprints
  tasks: Task[];
  burndownData: BurndownPoint[];
  createdAt: Date;
  updatedAt: Date;
}

export interface BurndownPoint {
  date: Date;
  remainingWork: number;
  idealRemaining: number;
}

export interface Project {
  id: string;
  name: string;
  description?: string;
  key: string; // project key like "ADV"
  type: 'software' | 'marketing' | 'research' | 'operations';
  status: 'planning' | 'active' | 'on_hold' | 'completed' | 'cancelled';
  visibility: 'public' | 'private' | 'team';
  leadId: string;
  leadName: string;
  teamMembers: ProjectMember[];
  startDate?: Date;
  endDate?: Date;
  budget?: number;
  spentBudget?: number;
  currency?: string;
  progress: number; // 0-100
  health: 'green' | 'yellow' | 'red';
  milestones: Milestone[];
  sprints: Sprint[];
  totalTasks: number;
  completedTasks: number;
  createdAt: Date;
  updatedAt: Date;
}

export interface ProjectMember {
  userId: string;
  userName: string;
  email: string;
  role: 'owner' | 'admin' | 'developer' | 'designer' | 'tester' | 'viewer';
  joinedAt: Date;
  allocation: number; // percentage of time allocated to project
  isActive: boolean;
}

export interface ResourceAllocation {
  id: string;
  userId: string;
  userName: string;
  email: string;
  role: string;
  department: string;
  totalCapacity: number; // hours per week
  allocations: ProjectAllocation[];
  availability: AvailabilityPeriod[];
  skills: Skill[];
  hourlyRate?: number;
  currency?: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface ProjectAllocation {
  projectId: string;
  projectName: string;
  allocation: number; // percentage or hours
  startDate: Date;
  endDate?: Date;
  role: string;
  isActive: boolean;
}

export interface AvailabilityPeriod {
  id: string;
  type: 'vacation' | 'sick_leave' | 'training' | 'conference' | 'unavailable';
  startDate: Date;
  endDate: Date;
  description?: string;
  isApproved: boolean;
  approvedBy?: string;
  createdAt: Date;
}

export interface Skill {
  id: string;
  name: string;
  category: string;
  level: 'beginner' | 'intermediate' | 'advanced' | 'expert';
  yearsOfExperience?: number;
  certifications?: string[];
}

export interface TimeEntry {
  id: string;
  userId: string;
  userName: string;
  taskId?: string;
  projectId?: string;
  description: string;
  hours: number;
  date: Date;
  billable: boolean;
  hourlyRate?: number;
  currency?: string;
  status: 'draft' | 'submitted' | 'approved' | 'rejected';
  approvedBy?: string;
  approvedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

export interface ProjectReport {
  id: string;
  projectId: string;
  type: 'status' | 'progress' | 'budget' | 'resource' | 'velocity' | 'burndown';
  title: string;
  period: DateRange;
  data: Record<string, any>;
  generatedBy: string;
  generatedAt: Date;
  format: 'json' | 'pdf' | 'excel' | 'csv';
  url?: string;
}

export interface ProjectDashboard {
  totalProjects: number;
  activeProjects: number;
  completedProjects: number;
  overdueTasks: number;
  totalTeamMembers: number;
  averageProjectHealth: number;
  budgetUtilization: number;
  velocityTrend: VelocityPoint[];
  upcomingDeadlines: Milestone[];
  resourceUtilization: ResourceUtilization[];
  recentActivity: ProjectActivity[];
  period: DateRange;
  lastUpdated: Date;
}

export interface VelocityPoint {
  sprintName: string;
  planned: number;
  completed: number;
  date: Date;
}

export interface ResourceUtilization {
  userId: string;
  userName: string;
  totalCapacity: number;
  allocated: number;
  utilization: number; // percentage
  overallocated: boolean;
}

export interface ProjectActivity {
  id: string;
  type: 'task_created' | 'task_completed' | 'milestone_reached' | 'sprint_started' | 'sprint_completed';
  description: string;
  userId: string;
  userName: string;
  projectId?: string;
  projectName?: string;
  taskId?: string;
  taskTitle?: string;
  createdAt: Date;
}

export interface KanbanBoard {
  id: string;
  name: string;
  projectId?: string;
  columns: KanbanColumn[];
  filters: KanbanFilters;
  createdAt: Date;
  updatedAt: Date;
}

export interface KanbanColumn {
  id: string;
  name: string;
  status: string;
  order: number;
  wipLimit?: number; // work in progress limit
  tasks: Task[];
  color?: string;
}

export interface KanbanFilters {
  assignee?: string[];
  priority?: string[];
  type?: string[];
  labels?: string[];
  milestone?: string;
  sprint?: string;
}

export interface DateRange {
  start: Date;
  end: Date;
}

// API Response types
export interface ProjectApiResponse<T> {
  data: T;
  success: boolean;
  message?: string;
  pagination?: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

export interface ProjectFilters {
  status?: string[];
  type?: string[];
  assignee?: string[];
  priority?: string[];
  milestone?: string[];
  sprint?: string[];
  dateRange?: DateRange;
  labels?: string[];
}

// Form types for creating/editing
export interface CreateTaskRequest {
  title: string;
  description?: string;
  type: 'feature' | 'bug' | 'improvement' | 'maintenance' | 'research' | 'documentation';
  priority: 'low' | 'medium' | 'high' | 'critical';
  assigneeId?: string;
  projectId?: string;
  milestoneId?: string;
  sprintId?: string;
  labels: string[];
  estimatedHours?: number;
  storyPoints?: number;
  dueDate?: Date;
  parentTaskId?: string;
}

export interface CreateMilestoneRequest {
  name: string;
  description?: string;
  projectId?: string;
  startDate: Date;
  dueDate: Date;
  budget?: number;
  currency?: string;
}

export interface CreateSprintRequest {
  name: string;
  goal?: string;
  projectId?: string;
  startDate: Date;
  endDate: Date;
  capacity: number;
}

export interface CreateProjectRequest {
  name: string;
  description?: string;
  key: string;
  type: 'software' | 'marketing' | 'research' | 'operations';
  visibility: 'public' | 'private' | 'team';
  leadId: string;
  startDate?: Date;
  endDate?: Date;
  budget?: number;
  currency?: string;
}

export interface UpdateResourceAllocationRequest {
  allocations: {
    projectId: string;
    allocation: number;
    startDate: Date;
    endDate?: Date;
    role: string;
  }[];
  availability: Omit<AvailabilityPeriod, 'id' | 'isApproved' | 'approvedBy' | 'createdAt'>[];
}

export interface CreateTimeEntryRequest {
  taskId?: string;
  projectId?: string;
  description: string;
  hours: number;
  date: Date;
  billable: boolean;
}

File: backend/app/models/project.py
from sqlalchemy import Column, Integer, String, Float, Boolean, DateTime, Text, JSON, ForeignKey, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from enum import Enum as PyEnum
from datetime import datetime
from typing import Optional, Dict, Any, List

from ..core.database import Base


class ProjectStatus(PyEnum):
    PLANNING = "planning"
    ACTIVE = "active"
    ON_HOLD = "on_hold"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class ProjectType(PyEnum):
    SOFTWARE = "software"
    MARKETING = "marketing"
    RESEARCH = "research"
    OPERATIONS = "operations"


class ProjectVisibility(PyEnum):
    PUBLIC = "public"
    PRIVATE = "private"
    TEAM = "team"


class TaskStatus(PyEnum):
    BACKLOG = "backlog"
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    TESTING = "testing"
    DONE = "done"
    CANCELLED = "cancelled"


class TaskType(PyEnum):
    FEATURE = "feature"
    BUG = "bug"
    IMPROVEMENT = "improvement"
    MAINTENANCE = "maintenance"
    RESEARCH = "research"
    DOCUMENTATION = "documentation"


class TaskPriority(PyEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class MilestoneStatus(PyEnum):
    PLANNING = "planning"
    ACTIVE = "active"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    OVERDUE = "overdue"


class SprintStatus(PyEnum):
    PLANNING = "planning"
    ACTIVE = "active"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class ProjectHealth(PyEnum):
    GREEN = "green"
    YELLOW = "yellow"
    RED = "red"


class Project(Base):
    __tablename__ = "projects"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False, index=True)
    description = Column(Text)
    key = Column(String(10), nullable=False, unique=True, index=True)  # Project key like "ADV"
    
    # Project configuration
    type = Column(Enum(ProjectType), nullable=False, default=ProjectType.SOFTWARE)
    status = Column(Enum(ProjectStatus), nullable=False, default=ProjectStatus.PLANNING)
    visibility = Column(Enum(ProjectVisibility), nullable=False, default=ProjectVisibility.TEAM)
    health = Column(Enum(ProjectHealth), nullable=False, default=ProjectHealth.GREEN)
    
    # Project leadership
    lead_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Timeline
    start_date = Column(DateTime(timezone=True), nullable=True)
    end_date = Column(DateTime(timezone=True), nullable=True)
    
    # Budget
    budget = Column(Float, nullable=True)
    spent_budget = Column(Float, nullable=False, default=0.0)
    currency = Column(String(3), nullable=False, default='RUB')
    
    # Progress tracking
    progress = Column(Float, nullable=False, default=0.0)  # 0-100
    total_tasks = Column(Integer, nullable=False, default=0)
    completed_tasks = Column(Integer, nullable=False, default=0)
    
    # Metadata
    settings = Column(JSON)  # Project-specific settings
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    lead = relationship("User", foreign_keys=[lead_id])
    members = relationship("ProjectMember", back_populates="project", cascade="all, delete-orphan")
    tasks = relationship("Task", back_populates="project", cascade="all, delete-orphan")
    milestones = relationship("Milestone", back_populates="project", cascade="all, delete-orphan")
    sprints = relationship("Sprint", back_populates="project", cascade="all, delete-orphan")
    time_entries = relationship("TimeEntry", back_populates="project", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Project(id={self.id}, name='{self.name}', key='{self.key}')>"


class ProjectMember(Base):
    __tablename__ = "project_members"

    id = Column(Integer, primary_key=True, index=True)
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Role and permissions
    role = Column(String(50), nullable=False, default='developer')  # owner, admin, developer, designer, tester, viewer
    allocation = Column(Float, nullable=False, default=100.0)  # Percentage of time allocated
    
    # Status
    is_active = Column(Boolean, nullable=False, default=True)
    joined_at = Column(DateTime(timezone=True), server_default=func.now())
    left_at = Column(DateTime(timezone=True), nullable=True)
    
    # Relationships
    project = relationship("Project", back_populates="members")
    user = relationship("User", foreign_keys=[user_id])

    def __repr__(self):
        return f"<ProjectMember(project_id={self.project_id}, user_id={self.user_id}, role='{self.role}')>"


class Task(Base):
    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(255), nullable=False, index=True)
    description = Column(Text)
    
    # Task classification
    type = Column(Enum(TaskType), nullable=False, default=TaskType.FEATURE)
    status = Column(Enum(TaskStatus), nullable=False, default=TaskStatus.BACKLOG)
    priority = Column(Enum(TaskPriority), nullable=False, default=TaskPriority.MEDIUM)
    
    # Assignment
    assignee_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    reporter_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Project relationships
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=True)
    milestone_id = Column(Integer, ForeignKey("milestones.id"), nullable=True)
    sprint_id = Column(Integer, ForeignKey("sprints.id"), nullable=True)
    
    # Task hierarchy
    parent_task_id = Column(Integer, ForeignKey("tasks.id"), nullable=True)
    
    # Estimation and tracking
    estimated_hours = Column(Float, nullable=True)
    actual_hours = Column(Float, nullable=False, default=0.0)
    story_points = Column(Integer, nullable=True)
    
    # Timeline
    due_date = Column(DateTime(timezone=True), nullable=True)
    start_date = Column(DateTime(timezone=True), nullable=True)
    completed_date = Column(DateTime(timezone=True), nullable=True)
    
    # Metadata
    labels = Column(JSON)  # List of labels/tags
    custom_fields = Column(JSON)  # Custom field values
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    assignee = relationship("User", foreign_keys=[assignee_id])
    reporter = relationship("User", foreign_keys=[reporter_id])
    project = relationship("Project", back_populates="tasks")
    milestone = relationship("Milestone", back_populates="tasks")
    sprint = relationship("Sprint", back_populates="tasks")
    parent_task = relationship("Task", remote_side=[id])
    subtasks = relationship("Task", back_populates="parent_task")
    comments = relationship("TaskComment", back_populates="task", cascade="all, delete-orphan")
    attachments = relationship("TaskAttachment", back_populates="task", cascade="all, delete-orphan")
    dependencies = relationship("TaskDependency", foreign_keys="TaskDependency.task_id", back_populates="task")
    time_entries = relationship("TimeEntry", back_populates="task", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Task(id={self.id}, title='{self.title}', status='{self.status}')>"


class TaskComment(Base):
    __tablename__ = "task_comments"

    id = Column(Integer, primary_key=True, index=True)
    task_id = Column(Integer, ForeignKey("tasks.id"), nullable=False)
    author_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    content = Column(Text, nullable=False)
    is_edited = Column(Boolean, nullable=False, default=False)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    task = relationship("Task", back_populates="comments")
    author = relationship("User", foreign_keys=[author_id])

    def __repr__(self):
        return f"<TaskComment(id={self.id}, task_id={self.task_id})>"


class TaskAttachment(Base):
    __tablename__ = "task_attachments"

    id = Column(Integer, primary_key=True, index=True)
    task_id = Column(Integer, ForeignKey("tasks.id"), nullable=False)
    
    file_name = Column(String(255), nullable=False)
    file_size = Column(Integer, nullable=False)
    file_type = Column(String(100), nullable=False)
    file_path = Column(String(500), nullable=False)
    
    uploaded_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    uploaded_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    task = relationship("Task", back_populates="attachments")
    uploader = relationship("User", foreign_keys=[uploaded_by])

    def __repr__(self):
        return f"<TaskAttachment(id={self.id}, file_name='{self.file_name}')>"


class TaskDependency(Base):
    __tablename__ = "task_dependencies"

    id = Column(Integer, primary_key=True, index=True)
    task_id = Column(Integer, ForeignKey("tasks.id"), nullable=False)
    depends_on_task_id = Column(Integer, ForeignKey("tasks.id"), nullable=False)
    
    type = Column(String(50), nullable=False, default='blocks')  # blocks, relates_to, duplicates, caused_by
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    task = relationship("Task", foreign_keys=[task_id], back_populates="dependencies")
    depends_on_task = relationship("Task", foreign_keys=[depends_on_task_id])

    def __repr__(self):
        return f"<TaskDependency(task_id={self.task_id}, depends_on={self.depends_on_task_id})>"


class Milestone(Base):
    __tablename__ = "milestones"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False, index=True)
    description = Column(Text)
    
    # Project relationship
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=True)
    
    # Status and timeline
    status = Column(Enum(MilestoneStatus), nullable=False, default=MilestoneStatus.PLANNING)
    start_date = Column(DateTime(timezone=True), nullable=False)
    due_date = Column(DateTime(timezone=True), nullable=False)
    completed_date = Column(DateTime(timezone=True), nullable=True)
    
    # Progress tracking
    progress = Column(Float, nullable=False, default=0.0)  # 0-100
    total_tasks = Column(Integer, nullable=False, default=0)
    completed_tasks = Column(Integer, nullable=False, default=0)
    total_story_points = Column(Integer, nullable=True)
    completed_story_points = Column(Integer, nullable=False, default=0)
    
    # Budget
    budget = Column(Float, nullable=True)
    spent_budget = Column(Float, nullable=False, default=0.0)
    currency = Column(String(3), nullable=False, default='RUB')
    
    # Owner
    owner_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    project = relationship("Project", back_populates="milestones")
    owner = relationship("User", foreign_keys=[owner_id])
    tasks = relationship("Task", back_populates="milestone")

    def __repr__(self):
        return f"<Milestone(id={self.id}, name='{self.name}', status='{self.status}')>"


class Sprint(Base):
    __tablename__ = "sprints"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False, index=True)
    goal = Column(Text)
    
    # Project relationship
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=True)
    
    # Status and timeline
    status = Column(Enum(SprintStatus), nullable=False, default=SprintStatus.PLANNING)
    start_date = Column(DateTime(timezone=True), nullable=False)
    end_date = Column(DateTime(timezone=True), nullable=False)
    
    # Capacity and commitment
    capacity = Column(Integer, nullable=False)  # Story points or hours
    commitment = Column(Integer, nullable=False, default=0)  # Planned story points or hours
    completed = Column(Integer, nullable=False, default=0)  # Actual completed story points or hours
    velocity = Column(Float, nullable=False, default=0.0)  # Average velocity from previous sprints
    
    # Burndown data
    burndown_data = Column(JSON)  # Daily burndown points
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    project = relationship("Project", back_populates="sprints")
    tasks = relationship("Task", back_populates="sprint")

    def __repr__(self):
        return f"<Sprint(id={self.id}, name='{self.name}', status='{self.status}')>"


class TimeEntry(Base):
    __tablename__ = "time_entries"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Work item relationships
    task_id = Column(Integer, ForeignKey("tasks.id"), nullable=True)
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=True)
    
    # Time tracking
    description = Column(Text, nullable=False)
    hours = Column(Float, nullable=False)
    date = Column(DateTime(timezone=True), nullable=False)
    
    # Billing
    billable = Column(Boolean, nullable=False, default=True)
    hourly_rate = Column(Float, nullable=True)
    currency = Column(String(3), nullable=False, default='RUB')
    
    # Approval workflow
    status = Column(String(20), nullable=False, default='draft')  # draft, submitted, approved, rejected
    approved_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    approved_at = Column(DateTime(timezone=True), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    user = relationship("User", foreign_keys=[user_id])
    task = relationship("Task", back_populates="time_entries")
    project = relationship("Project", back_populates="time_entries")
    approver = relationship("User", foreign_keys=[approved_by])

    def __repr__(self):
        return f"<TimeEntry(id={self.id}, user_id={self.user_id}, hours={self.hours})>"


class ResourceAllocation(Base):
    __tablename__ = "resource_allocations"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Capacity
    total_capacity = Column(Float, nullable=False, default=40.0)  # Hours per week
    
    # User information
    role = Column(String(100), nullable=False)
    department = Column(String(100), nullable=False)
    hourly_rate = Column(Float, nullable=True)
    currency = Column(String(3), nullable=False, default='RUB')
    
    # Skills and capabilities
    skills = Column(JSON)  # List of skills with levels
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    user = relationship("User", foreign_keys=[user_id])
    allocations = relationship("ProjectAllocation", back_populates="resource")
    availability = relationship("AvailabilityPeriod", back_populates="resource")

    def __repr__(self):
        return f"<ResourceAllocation(id={self.id}, user_id={self.user_id})>"


class ProjectAllocation(Base):
    __tablename__ = "project_allocations"

    id = Column(Integer, primary_key=True, index=True)
    resource_id = Column(Integer, ForeignKey("resource_allocations.id"), nullable=False)
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    
    # Allocation details
    allocation = Column(Float, nullable=False)  # Percentage or hours
    role = Column(String(100), nullable=False)
    start_date = Column(DateTime(timezone=True), nullable=False)
    end_date = Column(DateTime(timezone=True), nullable=True)
    is_active = Column(Boolean, nullable=False, default=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    resource = relationship("ResourceAllocation", back_populates="allocations")
    project = relationship("Project", foreign_keys=[project_id])

    def __repr__(self):
        return f"<ProjectAllocation(resource_id={self.resource_id}, project_id={self.project_id})>"


class AvailabilityPeriod(Base):
    __tablename__ = "availability_periods"

    id = Column(Integer, primary_key=True, index=True)
    resource_id = Column(Integer, ForeignKey("resource_allocations.id"), nullable=False)
    
    # Availability details
    type = Column(String(50), nullable=False)  # vacation, sick_leave, training, conference, unavailable
    start_date = Column(DateTime(timezone=True), nullable=False)
    end_date = Column(DateTime(timezone=True), nullable=False)
    description = Column(Text)
    
    # Approval
    is_approved = Column(Boolean, nullable=False, default=False)
    approved_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    resource = relationship("ResourceAllocation", back_populates="availability")
    approver = relationship("User", foreign_keys=[approved_by])

    def __repr__(self):
        return f"<AvailabilityPeriod(id={self.id}, type='{self.type}')>"